corehq/ex-submodules/casexml/apps/phone/tests/test_index_tree.py

```diff
@@ -0,0 +1,161 @@
+        self.assertFalse(child_id in tree.dependent_indices)
+
+    def test_prune_tiered_top_down(self):
+        [grandparent_id, parent_id, child_id] = ['grandparent', 'parent', 'child']
+        tree = IndexTree(live_indices={
+            child_id: [parent_id],
+            parent_id: [grandparent_id],
+        })
+        self.assertTrue(parent_id in tree.live_indices)
+        self.assertTrue(child_id in tree.live_indices)
+        self.assertFalse(parent_id in tree.dependent_indices)
+        self.assertFalse(child_id in tree.dependent_indices)
+
+        # this has no effect
+        tree.prune_case(grandparent_id)
+        self.assertTrue(parent_id in tree.live_indices)
+        self.assertTrue(child_id in tree.live_indices)
+        self.assertFalse(parent_id in tree.dependent_indices)
+        self.assertFalse(child_id in tree.dependent_indices)
+
+        # this should push the parent from live to dependent
+        tree.prune_case(parent_id)
+        self.assertFalse(parent_id in tree.live_indices)
+        self.assertTrue(parent_id in tree.dependent_indices)
+        self.assertTrue(child_id in tree.live_indices)
+        self.assertFalse(child_id in tree.dependent_indices)
+
+        # this should prune everything
+        tree.prune_case(child_id)
+        self.assertFalse(parent_id in tree.live_indices)
+        self.assertFalse(parent_id in tree.dependent_indices)
+        self.assertFalse(child_id in tree.live_indices)
+        self.assertFalse(child_id in tree.dependent_indices)
+
+    def test_prune_tiered_bottom_up(self):
+        [grandparent_id, parent_id, child_id] = ['grandparent', 'parent', 'child']
+        tree = IndexTree(live_indices={
+            child_id: [parent_id],
+            parent_id: [grandparent_id],
+        })
+        # just pruing the child should prune everything
+        tree.prune_case(child_id)
+        self.assertFalse(parent_id in tree.live_indices)
+        self.assertFalse(parent_id in tree.dependent_indices)
+        self.assertFalse(child_id in tree.live_indices)
+        self.assertFalse(child_id in tree.dependent_indices)
+
+    def test_prune_multiple_children(self):
+        [grandparent_id, parent_id, child_id_1, child_id_2] = ['rickard', 'ned', 'bran', 'arya']
+        tree = IndexTree(live_indices={
+            child_id_1: [parent_id],
+            child_id_2: [parent_id],
+            parent_id: [grandparent_id],
+        })
+        # just pruning one child should preserve the parent index
+        tree.prune_case(child_id_1)
+        self.assertTrue(parent_id in tree.live_indices)
+        self.assertFalse(parent_id in tree.dependent_indices)
+        self.assertFalse(child_id_1 in tree.live_indices)
+        self.assertFalse(child_id_1 in tree.dependent_indices)
+        self.assertTrue(child_id_2 in tree.live_indices)
+        self.assertFalse(child_id_2 in tree.dependent_indices)
+
+        # pruning the other one should wipe it
```

Comment:
possibly - actually stop reviewing this for now - i think i need to rework it (again) :(